File: board.py
from ship import ThreePieceShip, SmallL, BigL

class Board:

    """
    Water: 0
    ship: 1
    Miss: 3
    Hit: 4
    """
    WATER = 0
    SHIP = 1
    MISS = 3
    HIT = 4
    HOLE_SHIP_SUNK = 5

    def __init__(self) -> None:
        self.board = [[0 for i in range(10)] for j in range(10)]
        self.ship_dict = {}
        self.ship_coordinates = {}
    
    def print_board(self) -> None:
        for i in range(len(self.board)):
            print(self.board[i])

    def place_ship(self, coordinate = (5,5), ship = None) -> bool:
        """
        placement: list of coordinates to place the ship
        Returns: true if ship could be placed
        """
        coordinate_x, coordinate_y = coordinate
        placement = []
        for x, y in ship.offsets:
            placement_x = x + coordinate_x
            placement_y = y + coordinate_y
            placement.append((placement_x, placement_y))

        if self.check_legal_ship_placement(placement):
            for x, y in placement:
                self.board[y][x] = self.SHIP
        else:
            return False

        for ship_coordinate in placement:
            self.ship_dict[ship_coordinate] = ship
        
        self.ship_coordinates[ship] = placement
        ship.placed = True
        return True

    def place_missile(self, coordinate) -> dict: # må returnere en dict med coordinat og verdi til coordinat
        affected_coordinates = {}
        x, y = coordinate
        if self.board[y][x] > 1:
            return affected_coordinates
        self.board[y][x] += 3
        if coordinate in self.ship_dict.keys():
            self.ship_dict[coordinate].damage_ship()
            if self.ship_dict[coordinate].is_sunk():
                return self.sink_ship(self.ship_dict[coordinate])
        return {coordinate : self.board[y][x]}

    def sink_ship(self, ship): # TODO
        affected_coordinates = {}
        for x, y in self.ship_coordinates[ship]:
            self.board[y][x] = self.HOLE_SHIP_SUNK
            affected_coordinates[(x,y)] = self.HOLE_SHIP_SUNK
        return affected_coordinates

    def check_legal_ship_placement(self, placement):
        for x, y in placement:
            if x < 0 or y < 0:
                print("SHIP OUT OF BOUND")
                return False
            if self.board[y][x] != self.WATER:
                print("SHIP COLLITION WITH OTHER SHIP")
                return False
        return True


    def check_game_over(self):
        for y in range(len(self.board)):
            for x in range(len(self.board[y])):
                if self.board[y][x] == self.SHIP:
                    return False
        
        return True

    
        




if __name__ == "__main__":
    b = Board()
    s = BigL()
    s.rotate_right()
    p = b.place_ship(1, (5,5), s)
    b.print_board()
    while not b.check_game_over():
        x_input = int(input("x: "))
        y_input = int(input("y: "))
        b.place_missile((x_input, y_input))
        b.print_board() 

File: server/game_logic/board.js
export class Board {

    constructor(){
        this.WATER = 0
        this.SHIP = 1
        this.MISS = 3
        this.HIT = 4
        this.HOLE_SHIP_SUNK = 5
        this.board = this.initBoard();
        this.shipDict = new Map()
        this.shipCoordinates = new Map()
    }

    printBoard(){
        for (var i = 0; i < this.board.length; i++){
            console.log(this.board[i]);
        }
    }

    placeShip(coordinate, ship){
        var coordinate_x = coordinate.x;
        var coordinate_y = coordinate.y;
        var placement = [];
        for (var i = 0; i < ship.offsets.length; i++){
            var placement_x = ship.offsets[i].x + coordinate_x;
            var placement_y = ship.offsets[i].y + coordinate_y;
            placement.push({"x":placement_x, "y":placement_y});
        }

        if (!ship.placed && this.checkLegalShipPlacement(placement)){
            for (var i = 0; i < placement.length; i++){
                this.board[placement[i].y][placement[i].x] = this.SHIP;
            }
        } else {
            return false
        }
        
        for (var i = 0; i < placement.length; i++){
            var key = placement[i].x + "," + placement[i].y;
            this.shipDict.set(key, ship); 
        }
        this.shipCoordinates.set(ship, placement);
        ship.placed = true;
        return true
    }

    placeMissile(coordinate){
        var affectedCoordinates = new Map();
        var x = coordinate.x;
        var y = coordinate.y;
        if (this.board[y][x] > 1){
            return affectedCoordinates;
        }
        this.board[y][x] += 3;
        var key = x + "," + y;
        if (this.shipDict.has(key)){
            this.shipDict.get(key).damageShip();
            if (this.shipDict.get(key).isSunk()){
                return this.sinkShip(this.shipDict.get(key));
            }
        }
        affectedCoordinates.set(key, this.board[y][x]);
        return affectedCoordinates;
    }

    sinkShip(ship){
        var affectedCoordinates = new Map();
        var coordinates = this.shipCoordinates.get(ship);
        for (var i = 0; i < coordinates.length; i++){
            var x = coordinates[i].x;
            var y = coordinates[i].y;
            this.board[y][x] = this.HOLE_SHIP_SUNK;
            var key = x + "," + y;
            affectedCoordinates.set(key, this.HOLE_SHIP_SUNK);
        }
        return affectedCoordinates;
    }

    checkLegalShipPlacement(placement){
        for (var i = 0; i < placement.length; i++){
            var x = placement[i].x;
            var y = placement[i].y;
            if (x < 0 || y < 0){
                console.log("SHIP out of bound");
                return false
            }
            if (this.board[y][x] !== this.WATER){
                console.log("Ship collition with other ship");
                return false
            }
        }
        return true
    }

    checkGameOver(){
        for (var y = 0; y < this.board.length; y++){
            for (var x = 0; x < this.board[y].length; x++){
                if (this.board[y][x] === this.SHIP){
                    return false
                }
            }
        }
        return true
    }

    initBoard(){
        var board = []
        for (var i = 0; i < 10; i++){
            var row = []
            for (var j = 0; j < 10; j++){
                row.push(0)
            }
            board.push(row)
        }
        return board
    }


}
